slash/log.py

```diff
@@ -193,85 +191,82 @@ def _should_delete_log(self, result):
-                handler = logbook.GZIPCompressionHandler
-                filename += ".gz"
+                log_path += ".gz"
             elif config.root.log.compression.algorithm == "brotli":
-                handler = logbook.BrotliCompressionHandler
-                filename += ".brotli"
+                log_path += ".brotli"
             else:
                 raise InvalidConfiguraion("Unsupported compression method: {}".format(config.root.log.compression.algorithm))
+        return log_path
+
+    def _create_log_file_handler(self, log_path, bubble=False, filter=_slash_logs_filter, use_compression=False, use_rotation=False):
+        kwargs = {"bubble": bubble, "filter": filter}
+        if use_compression:
+            if config.root.log.compression.algorithm == "gzip":
+                handler_class = logbook.GZIPCompressionHandler
+            elif config.root.log.compression.algorithm == "brotli":
+                handler_class = logbook.BrotliCompressionHandler
         elif use_rotation:
             kwargs.update({"max_size": 4*1024**2, "backup_count": 1})
-            handler = logbook.RotatingFileHandler
+            handler_class = logbook.RotatingFileHandler
         elif config.root.log.colorize:
-            handler = ColorizedFileHandler
+            handler_class = ColorizedFileHandler
         else:
-            handler = logbook.FileHandler
-        return handler(filename, **kwargs)
+            handler_class = logbook.FileHandler
+        return handler_class(log_path, **kwargs)
+
+    @contextmanager
+    def _log_file_handler_context(self, subpath, symlink, bubble=False, filter=_slash_logs_filter, use_compression=False, use_rotation=False):
+        if subpath is None or config.root.log.root is None:
+            yield NoopHandler() if bubble else logbook.NullHandler(filter=filter)
+        else:
+            log_path = self._get_log_file_path(subpath, use_compression)
+            handler = self._log_path_to_handler.get(log_path, None)
+            if handler is not None:
+                yield handler
+            else:
+                result = context.result
+                ensure_containing_directory(log_path)
+                if symlink:
+                    self._try_create_symlink(log_path, symlink)
+                with closing(self._create_log_file_handler(log_path, bubble=bubble, use_compression=use_compression, \
+                                                           use_rotation=use_rotation, filter=filter)) as handler:
+                    self._log_path_to_handler[log_path] = handler
+                    self._set_formatting(handler, config.root.log.format)
+                    yield handler
+                self._log_path_to_handler[log_path] = None
+                hooks.log_file_closed(path=log_path, result=result)  # pylint: disable=no-member
+                if config.root.log.cleanup.enabled and self._should_delete_log(result):
+                    os.remove(log_path)
```

Comment:
Not entirely related to this PR, but I think failure to remove a log should just add an error to the session and not fail it completely...